File: solutions/typescript/transpose/1/transpose.ts
/**
 * Transpone una matriz de strings (filas se convierten en columnas).
 * Maneja filas de longitudes desiguales aplicando las reglas de relleno específicas.
 * * @param input - Arreglo de strings que representan las líneas originales.
 * @returns Arreglo de strings transpuesto.
 */
export function transpose(input: string[]): string[] {
  if (input.length === 0) {
    return [];
  }

  // 1. Encontrar la longitud de la línea más larga para saber cuántas filas de salida tendremos
  const maxCols = Math.max(...input.map(row => row.length));
  const result: string[] = [];

  // 2. Iterar a través de cada columna de la matriz original (que será una nueva fila)
  for (let c = 0; c < maxCols; c++) {
    let currentNewRow = '';

    // Iterar a través de cada fila de la matriz original
    for (let r = 0; r < input.length; r++) {
      const char = input[r][c];

      if (char !== undefined) {
        currentNewRow += char;
      } else {
        // Si el carácter no existe en la fila 'r', verificamos si alguna de las
        // FILAS SIGUIENTES (r + 1 en adelante) es lo suficientemente larga como para alcanzar esta columna 'c'.
        const hasLongerRowAhead = input.slice(r + 1).some(nextRow => nextRow.length > c);

        if (hasLongerRowAhead) {
          currentNewRow += ' ';
        } else {
          // Si ninguna fila posterior necesita esta columna, detenemos el procesamiento
          // de la fila actual para no rellenar espacios a la derecha de forma innecesaria.
          break;
        }
      }
    }

    result.push(currentNewRow);
  }

  return result;
}

File: solutions/ruby/transpose/1/transpose.rb
class Transpose
  def self.transpose(input)
    # 1. Convertimos el input en un arreglo de filas
    lines = input.split("\n")
    return "" if lines.empty?

    # 2. Calculamos la longitud máxima de las filas
    max_length = lines.map(&:length).max
    
    # 3. Transponemos recorriendo cada índice de columna posible
    result = (0...max_length).map do |col_idx|
      # Para cada columna, iteramos sobre las filas originales
      lines.each_with_index.map do |line, row_idx|
        # Buscamos la longitud máxima que alcanzan las filas que vienen después
        # Esto nos dirá si debemos rellenar con un espacio o dejar vacío
        max_future_length = lines[row_idx..].map(&:length).max
        
        if col_idx < line.length
          line[col_idx]
        elsif col_idx < max_future_length
          " "
        else
          nil
        end
      end.join
    end

    # 4. Unimos las nuevas filas con saltos de línea
    result.join("\n")
  end
end